broker/bam/src/ba_best.cc

```diff
@@ -96,29 +97,54 @@ state ba_best::get_state_soft() const {
  *
  *  @param[in] impact Impact information.
  */
-bool ba_best::_apply_impact(kpi* kpi_ptr __attribute__((unused)),
+void ba_best::_apply_impact(kpi* kpi_ptr [[maybe_unused]],
                             ba::impact_info& impact) {
-  const std::array<short, 5> order{0, 3, 4, 2, 1};
-
-  auto is_state_better = [&](short current_state, short new_state) -> bool {
+  auto is_state_better = [](short current_state, short new_state) -> bool {
+    const std::array<short, 5> order{0, 3, 4, 2, 1};
     assert((unsigned int)current_state < order.size());
     assert((unsigned int)new_state < order.size());
     return order[new_state] < order[current_state];
   };
 
   if (_dt_behaviour == configuration::ba::dt_ignore_kpi && impact.in_downtime)
-    return false;
+    return;
 
-  bool retval = false;
-  if (is_state_better(_computed_soft_state, impact.soft_impact.get_state())) {
+  if (is_state_better(_computed_soft_state, impact.soft_impact.get_state()))
     _computed_soft_state = impact.soft_impact.get_state();
-    retval = true;
-  }
-  if (is_state_better(_computed_hard_state, impact.hard_impact.get_state())) {
+  if (is_state_better(_computed_hard_state, impact.hard_impact.get_state()))
     _computed_hard_state = impact.hard_impact.get_state();
-    retval = true;
+}
+
+/**
+ *  Apply some impact. This method is more complete than _apply_impact().
+ *  It takes as argument a child kpi and its impact. This child is already
+ *  known, so its previous impact is replaced by the new one.
+ *  In other words, this method makes almost the same work as _unapply_impact()
+ *  and the _apply_impact() ; the difference is that it returns true if the BA
+ *  really changed.
+ *
+ *  @param[in] impact Impact information.
+ *  @return True if the BA changes, False otherwise.
+ */
+bool ba_best::_apply_changes(kpi* child,
+                             const impact_values& new_hard_impact,
+                             const impact_values& new_soft_impact,
+                             bool in_downtime) {
+  state previous_state = _computed_hard_state;
+
+  _computed_soft_state = _computed_hard_state = state_critical;
+
+  // We recompute all impacts...
```

Comment:
why recompute all impacts on change of only one kpi?

Reply:
Not all, just the one of the BA. It is of type BEST. So we have to all the children to get the best one, haven't we?